midend/interpreter.cpp

```diff
@@ -731,6 +731,20 @@ void ExpressionEvaluator::postorder(const IR::Operation_Unary* expression) {
         DoConstantFolding cf(refMap, typeMap);
         cf.setCalledBy(this);
         auto result = expression->apply(cf);
+        if (!result->is<IR::Constant>()) {
```

Comment:
This is not how this should be done.
If you want to handle casts, use the postorder function for casts.
Also, you still have not explained why this case can ever happen. 
After you call constantFolding one of two things should be true
- result is a constant, and then this code is never executed
- the cast argument is not a constant, and then you hit the BUG case below
So I think that all this code does actually nothing.